Description: Added new C file (scrobbler.c) to compilation process
Author: Nanakos Chrysostomos <nanakos@wired-net.gr>

--- mpg321-0.2.11.orig/Makefile.am
+++ mpg321-0.2.11/Makefile.am
@@ -12,7 +12,8 @@ mpg321_SOURCES = \
 	remote.c \
 	ao.c \
 	options.c \
-	auth.c
+	auth.c \
+	scrobbler.c
 
 SUBDIRS = 
 EXTRA_DIST = README.remote HACKING BUGS mpg321.sgml mpg321.1 $(srcdir)/debian/*
--- mpg321-0.2.11.orig/Makefile.in
+++ mpg321-0.2.11/Makefile.in
@@ -57,7 +57,8 @@ binPROGRAMS_INSTALL = $(INSTALL_PROGRAM)
 PROGRAMS = $(bin_PROGRAMS)
 am_mpg321_OBJECTS = mpg321.$(OBJEXT) mad.$(OBJEXT) playlist.$(OBJEXT) \
 	network.$(OBJEXT) getopt.$(OBJEXT) getopt1.$(OBJEXT) \
-	remote.$(OBJEXT) ao.$(OBJEXT) options.$(OBJEXT) auth.${OBJEXT}
+	remote.$(OBJEXT) ao.$(OBJEXT) options.$(OBJEXT) auth.${OBJEXT} \
+	scrobbler.${OBJEXT}
 mpg321_OBJECTS = $(am_mpg321_OBJECTS)
 mpg321_LDADD = $(LDADD)
 DEFAULT_INCLUDES = -I. -I$(srcdir) -I.
@@ -180,7 +181,8 @@ mpg321_SOURCES = \
 	remote.c \
 	ao.c \
 	options.c \
-	auth.c
+	auth.c \
+	scrobbler.c
 
 SUBDIRS = 
 EXTRA_DIST = README.remote HACKING BUGS mpg321.sgml mpg321.1 $(srcdir)/debian/*
